cpp/src/io/json/read_json.cu

```diff
@@ -159,48 +182,84 @@ bool should_load_whole_source(json_reader_options const& opts, size_t source_siz
  * @param sources Data sources to read from
  * @param reader_opts JSON reader options with range offset and range size
  * @param stream CUDA stream used for device memory operations and kernel launches
- * @return Byte range for parsing
+ * @returns Data source owning buffer enclosing the bytes read
  */
-auto get_record_range_raw_input(host_span<std::unique_ptr<datasource>> sources,
-                                json_reader_options const& reader_opts,
-                                rmm::cuda_stream_view stream)
+datasource::owning_buffer<rmm::device_uvector<char>> get_record_range_raw_input(
+  host_span<std::unique_ptr<datasource>> sources,
+  json_reader_options const& reader_opts,
+  rmm::cuda_stream_view stream)
 {
-  auto buffer = ingest_raw_input(sources,
-                                 reader_opts.get_compression(),
-                                 reader_opts.get_byte_range_offset(),
-                                 reader_opts.get_byte_range_size(),
-                                 stream);
-  if (should_load_whole_source(reader_opts, sources[0]->size())) return buffer;
-  auto first_delim_pos =
-    reader_opts.get_byte_range_offset() == 0 ? 0 : find_first_delimiter(buffer, '\n', stream);
+  auto geometric_mean = [](double a, double b) { return std::pow(a * b, 0.5); };
+
+  size_t const total_source_size            = sources_size(sources, 0, 0);
+  auto constexpr num_delimiter_chars        = 1;
+  auto const num_extra_delimiters           = num_delimiter_chars * (sources.size() - 1);
+  size_t chunk_size                         = reader_opts.get_byte_range_size();
+  size_t chunk_offset                       = reader_opts.get_byte_range_offset();
+  compression_type const reader_compression = reader_opts.get_compression();
+
+  // Some magic numbers
+  constexpr int num_subchunks            = 10;  // per chunk_size
+  constexpr size_t min_subchunk_size     = 10000;
+  constexpr int num_subchunks_prealloced = 3;
+  constexpr int compression_ratio        = 4;
+  /*
+   * NOTE: heuristic for choosing subchunk size: geometric mean of minimum subchunk size (set to
+   * 10kb) and the byte range size
+   */
+  size_t size_per_subchunk = geometric_mean(
+    std::ceil((double)reader_opts.get_byte_range_size() / num_subchunks), min_subchunk_size);
```

Comment:
how does this work when byte range is not used? AFAIK get_byte_range_size returns zero in that case
EDIT: maybe we can use chunk_size after it's updated in the next branch.